### PR TITLE
Add method to check if job manager is correct

### PR DESCRIPTION
**Pull request by dankwart-de**
Correct means, correct on this submission host.
Different checks for different managers apply.

### CODE REVIEW

**Comment by dankwart-de** [comment]
First draft for this. What do you think?

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy [hunk 1]
```diff
@@ -71,6 +71,18 @@ class LSFJobManager extends AbstractLSFJobManager {
         super(executionService, parms)
     }
 
+    @Override
+    boolean checkCompatibilityToHost() {
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: The name of this method does not express its function. "compatibility" can be anything (Windows? Linux? Right distribution? LSF Version? etc.).

Please change it to e.g. `submissionHostCanSubmit` or similar.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy [hunk 2]
```diff
@@ -71,6 +71,18 @@ class LSFJobManager extends AbstractLSFJobManager {
...
+        def result = executionService.execute("which bsub; bsub -h 2>&1 >/dev/null | grep LSF > /dev/null && echo LSF || echo NONLSF")
+
+        // The result should now have two lines, e.g.:
+        //   /opt/lsf/10.1/linux3.10-glibc2.17-x86_64/bin/qsub
+        //   LSF
+        return result.resultLines.size() == 2 &&
+                result.resultLines[0].contains("lsf") &&
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: This condition is incorrect. The fact that 'lsf' is in our `bsub` path will certainly not hold other clusters.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy [hunk 3]
```diff
@@ -71,6 +71,18 @@ class LSFJobManager extends AbstractLSFJobManager {
...
+                result.resultLines[1] == "LSF"
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: Not sure. What about the newline?

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/rest/LSFRestJobManager.groovy [hunk 4]
```diff
@@ -61,7 +62,11 @@ class LSFRestJobManager extends AbstractLSFJobManager {
         this.restExecutionService = restExecutionService as RestExecutionService
     }
 
-    /**
+    @Override
+    boolean checkCompatibilityToHost() {
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/rest/LSFRestJobManager.groovy]: Why not check here that the REST endpoint indeed speaks LSF REST? With a not-implemented exception we will certainly need special try-catch in client code.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 5]
```diff
@@ -36,6 +36,24 @@ class PBSJobManager extends GridEngineBasedJobManager<PBSCommand> {
         super(executionService, parms)
     }
 
+    @Override
+    boolean checkCompatibilityToHost() {
+        // PBS tests are non trivial. qsub and co are not very helpful, as they don't feature any useful messages.
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: Also note that some cluster managers (LSF, SGE) or even add-on packages (e.g. for Slurm) provide `qsub`-named wrappers for their CLI APIs. I cannot tell whether these are true `qsub` compatible. Similar for other commands that we need (`qstat`).

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 6]
```diff
@@ -36,6 +36,24 @@ class PBSJobManager extends GridEngineBasedJobManager<PBSCommand> {
...
+        // not contain "torque".
+        String cmd = "which qsub;"
+        cmd += "ps -u root | grep trqauthd >/dev/null && echo PBS || NONPBS" // Append fast check for running torque auth daemon.
+        def result = executionService.execute(cmd)
+
+        // We should have two lines:
+        //   /opt/torque/bin/qsub
+        //   PBS / NONPBS
+        return result.resultLines.size() == 2 &&
+                result.resultLines[0].contains("torque") &&
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: As before this is not portable.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 7]
```diff
@@ -36,6 +36,24 @@ class PBSJobManager extends GridEngineBasedJobManager<PBSCommand> {
         super(executionService, parms)
     }
 
+    @Override
+    boolean checkCompatibilityToHost() {
+        // PBS tests are non trivial. qsub and co are not very helpful, as they don't feature any useful messages.
+        // As we support PBS Torque, we'll check installation folders and for a specific started service.
+        // Checking directories of course can miss, that Torque is installed in a different directory which does
+        // not contain "torque".
+        String cmd = "which qsub;"
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: Maybe you should also check something like the version? That may give you much more detailed information.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/sge/SGEJobManager.groovy [hunk 8]
```diff
@@ -36,6 +36,16 @@ class SGEJobManager extends GridEngineBasedJobManager<SGECommand> {
...
     }
 
+    @Override
+    boolean checkCompatibilityToHost() {
+        // For GridEngine, we check the installation folder of qsub. Should contain gridengine
+        // GridEngine does print some version code when you use -help option, but I am not sure, how this changes.
+
+        String cmd = "readlink -f `which qsub` | grep gridengine > /dev/null && echo GE || echo NONGE"
+        def result = executionService.execute(cmd)
+        return result.firstLine.contains("gridengine")
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/sge/SGEJobManager.groovy]: Dito. Not portable.

**Comment by dankwart-de** [comment]
I'll close it now. Let's discuss if we keep the code and if we try a new attempt.

**Pull request closed without merging**